docs/debugging-testing/testing/unit_testing.md

```diff
@@ -1,303 +1,363 @@
+1. Download CMake binaries from https://cmake.org/download/, and run the installer.
+1. Download Python 2.7 or Python 3 from https://www.python.org/getit/, and run the installer.
+1. Add MinGW, CMake and Python into system PATH.
+1. Install the addition Python package `gcovr` for code coverage report:
+    ```
+    pip install "gcovr>=4.1"
+    ```
+
+## Test code structure
+
+Unit tests are located in the `tests/UNITTESTS` subdirectory of each library. We recommend that unit test files use an identical directory path as the file under test. This makes it easier to find unit tests for a particular class or a module. For example, if the file you are testing is `some/example/path/ClassName.cpp`, then all the test files are in the `UNITTESTS/some/example/path/ClassName` directory. Each test suite needs to have its own `CMakeLists.txt` file for test CMake configuration.
+
+All the stub source files are built in a stub CMake library targets (e.g `mbed-stubs-rtos`) and linked to the `mbed-stubs` CMake target. The CMake target of the library unit under test is expected to link with the required stub libraries or `mbed-stubs` if it requires multiple stub libraries.
+
+The new stub file should follow the naming convention `<CLASS_NAME>_stub.cpp` for the source file and `<CLASS_NAME>_stub.h` for the header file. They should be built as part of their respective stub CMake library. Alternatively, create a stub library if `<CLASS_NAME>_stub` is an implementation for an external source not part of Mbed OS.
+
+All the Mbed OS header files are built with CMake `INTERFACE` libraries (e.g `mbed-headers-platform`). Stubbed header files reside in the `UNITTESTS/target_h` and are built with the `mbed-headers-base` CMake library. All CMake libraries containing header files are linked with `mbed-headers`. The CMake target of the library unit under test is expected to link with the required header file libraries or `mbed-headers` in case of requiring multiple header libraries. 
+
+All the stub libraries and header libraries are defined under `UNITTESTS/stubs/` directory.
+
+Libraries for fakes are under the `UNITTESTS/fakes` directory. These provide mock implementations that are meant to replace the stub version that does nothing. Usually, these will replace the header files as well as the source files and cannot be used together with their stub equivalents.
+
+The upcoming [Example](#Example) section describes more in detail with writing a unit test for a dummy example class (Unit Under Test), creating example header library, stub example interface library and target_link_libraries into existing mbed-stubs, mbed-headers and also in the unit test CMake configuration file.
+
+### Test discovery
+
+Registering unit tests to run happens automatically, and the test runner handles registration. However, test suites do not automatically build. 

```

Comment:
Automatically when? At CMake configure time? Build system generation time? What does "registering" mean in this context?

Reply:
`CMake configuration time` and `registering` refers to `add_test` call in every unit test CMake

Reply:
Should we add some of that information to the document?